File: wlr/SPX.Entities/EntityManager.cs
using System;
using System.Collections;
using System.Collections.Generic;
using System.Linq;
using SPX.Core;
using Microsoft.Xna.Framework.Graphics;

namespace SPX.Entities
{
    public static class EntityManager
    {
        private static List<Entity> _contents = new List<Entity>();
        private static Dictionary<Point2,List<Entity>> _gridContents = new Dictionary<Point2, List<Entity>>();

        public static Entity AddObject(Entity Entity)
        {
            Entity.LoadContent();
            _contents.Add(Entity);
            AddToGrid(Entity);
            return Entity;
        }

        public static Entity GetObject(int type)
        {
            if (_contents != null)
            {
                return _contents.FirstOrDefault(item => item.GetEntityType() == type);
            }
            return null;
        }

        private static readonly List<Entity> gopResults = new List<Entity>();
        public static List<Entity> GetEntities(int type,Point2 target)
        {
            if (_contents != null)
            {
                goResults.Clear();
                goResults.AddRange(GetObjects(type));
                gopResults.Clear();
                for (int ii = 0; ii < goResults.Count(); ii++)
                {
                    if (goResults[ii].Contains(target))
                    {
                        gopResults.Add(goResults[ii]);
                    }
                }
                return gopResults;
            }
            return null;
        }

        private static readonly List<Entity> goResults = new List<Entity>();
        public static List<Entity> GetObjects(int type)
        {
            goResults.Clear();
            for (int ii = 0; ii < _contents.Count(); ii++)
            {
                if (_contents[ii].GetEntityType() == type)
                {
                    goResults.Add(_contents[ii]);
                }
            }
            return goResults;
        }

        //CT Accessors
        public static IActor GetActor(int type)
        {
            return _contents != null ? _contents.Where(o => o.GetEntityType() == EntityType.ACTOR).Cast<IActor>().FirstOrDefault(creature => creature.GetActorType() == type) : null;
        }

        private static List<IActor> creatures = new List<IActor>();
        public static List<IActor> GetActors(int type)
        {
            creatures.Clear();
            if (type != ActorType.NONPLAYER)
            {
                foreach (var elem in _contents)
                {
                    if (elem.GetEntityType() == EntityType.ACTOR)
                    {
                        if (((IActor)elem).GetActorType() == type)
                        {
                            creatures.Add(((IActor)elem));
                        }
                    }
                }
            }
            else
            {
                foreach (var elem in _contents)
                {
                    if (elem.GetEntityType() == EntityType.ACTOR)
                    {
                        if (((IActor)elem).GetActorType() != ActorType.PLAYER)
                        {
                            creatures.Add(((IActor)elem));
                        }
                    }
                }
            }
            return creatures;
        }

        private static IActor _nextResult;
        public static List<IActor> GetActorsAt(Point2 target,int actorType = -1)
        {
            creatures.Clear();
            foreach (var elem in _gridContents[target])
            {
                if (elem.GetEntityType() == EntityType.ACTOR)
                {
                    _nextResult = (IActor)elem;
                    if (actorType == -1 
                        || _nextResult.GetActorType() == actorType
                        || (actorType == ActorType.NONPLAYER && _nextResult.GetActorType() != ActorType.PLAYER))
                    {
                        creatures.Add(_nextResult);
                    }
                }
            }
            return creatures;
        }

        private static List<IActor> creatures2 = new List<IActor>();
        public static List<IActor> GetActorsSurrounding(Point2 target,int distance)
        {
            creatures2.Clear();
            for (int ii = -distance; ii < distance+1; ii++)
            {
                for (int jj = -distance; jj < distance+1; jj++)
                {
                    if (ii != 0 || jj != 0)
                    {
                        foreach (var creature in GetActorsAt(target.Add(new Point2(ii, jj))))
                        {
                            creatures2.Add(creature);
                        }
                    }
                }
            }
            return creatures2;
        }

        public static bool IsLocationBlocked(Point2 location)
        {
            foreach(Entity elem in _gridContents[location])
            {
                if(elem.IsBlocking())
                {
                    return true;
                }
            }
            return false;
        }
        public static IEnumerable<Entity> GetObjectsToCache()
        {
            return _contents.Where(o => o.GetEntityType() != EntityType.FLOOR);
        }

        public static IActor GetNearestPlayer(Entity target)
        {
            var closest = GetActors(ActorType.PLAYER).FirstOrDefault() as Entity;
            foreach (var player in GetActors(ActorType.PLAYER).Select(x=>x as Entity))
            {
                if (HitTest.GetDistanceSquare(target, player) < HitTest.GetDistanceSquare(target, closest))
                {
                    closest = player;
                }
            }
            return closest as IActor;
        }

        public static IActor GetNearestPlayer(IActor target)
        {
            return (IActor)GetNearestPlayer((Entity)target);
        }

        public static void AddObjects(IEnumerable<Entity> cache)
        {
            _contents.AddRange(cache);
        }

        public static bool AnyContains(Point2 target, int type)
        {
            return _gridContents[target].Any(o => o.GetEntityType() == type);
        }

        public static void RemoveObject(Entity target)
        {
            _contents.Remove(target);
        }

        public static void Clear()
        {
            _contents = new List<Entity>();
            _gridContents = new Dictionary<Point2, List<Entity>>();
        }

        public static void Reset()
        {
            _contents = new List<Entity>();
            _gridContents = new Dictionary<Point2, List<Entity>>();
            creatures.Clear();
            LoadContent();
        }

        public static void Update()
        {   
            for (var ii = 0; ii < _contents.Count; ii++)
            {
                if(ii>=_contents.Count)
                {
                    return;
                }
                if (!_contents[ii].IsActive())
                {
                    _gridContents[_contents[ii].GetLocation()].Remove(_contents[ii]);
                    _contents.Remove(_contents[ii]);
                    ii--;
                    continue;
                }                
                _contents[ii].Update();
            }
        }

        public static void Draw()
        {
            if (XnaManager.Renderer != null)
            {
                foreach (var component in _contents)
                {
                    component.Draw();
                }
            }
        }

        public static void LoadContent()
        {
            if (XnaManager.Renderer != null)
            {
                foreach (var component in _contents)
                {
                    component.LoadContent();
                }
            }
        }

        private static void AddToGrid(Entity Entity)
        {
            if (!_gridContents.ContainsKey(Entity.GetLocation()))
            {
                _gridContents.Add(Entity.GetLocation(), new List<Entity>() { Entity });
            }
            else
            {
                _gridContents[Entity.GetLocation()].Add(Entity);
            }
        }

        public static void UpdateGridLocation(Entity Entity, Point2 oldLocation)
        {
            if(_gridContents!=null && oldLocation !=null)
            {
                _gridContents[oldLocation].Remove(Entity);
                AddToGrid(Entity);
            }
        }

        private static List<IActor> _players = new List<IActor>();
        public static IEnumerable<IActor> GetPlayers()
        {
            _players.Clear();
            foreach (var tile in _contents)
            {
                if (tile.GetEntityType() == EntityType.ACTOR && (tile as IActor).GetActorType() == ActorType.PLAYER)
                {
                    _players.Add((tile as IActor));
                }
            }
            return _players;
        }

        //Returns a space that has no creatures on it
        private static IEnumerable<Point2> emptyLocations;
        public static Point2 GetEmptyLocation()
        {
            emptyLocations = _gridContents.Where(gc =>
                !gc.Value.Any(e => e.GetEntityType() == EntityType.ACTOR) &&
                gc.Key.GridX > 0 && gc.Key.GridY > 0 &&
                gc.Key.GridX < GameManager.TileMapWidth - 1 && gc.Key.GridY < GameManager.TileMapHeight - 1).
                Select(gc => gc.Key);
            return emptyLocations.ElementAt(RNG.Rand.Next(0, emptyLocations.Count()));
        }

        //Deprecated methods
        public static ICollection<Entity> GetEntitiesToCache()
        {
            return _contents.Where(c => c.GetEntityType() != EntityType.FLOOR).ToList();
        }
        public static IActor GetTouchingCreature(Entity entity)
        {
            return _contents.Where(c=>c.GetEntityType() == EntityType.ACTOR).FirstOrDefault(c => c.Contains(entity.GetLocation())) as IActor;
        }
    }
}

File: proj/src/spx/entities/EntityManager.java
package spx.entities;

import spx.bridge.ActorType;
import spx.bridge.EntityType;
import spx.core.Point2;
import spx.core.RNG;
import spx.core.Settings;
import spx.core.SpxManager;

import java.util.ArrayList;
import java.util.HashMap;
import java.util.List;

public class EntityManager {
    private static List<IEntity> _contents = new ArrayList<>();
    private static HashMap<Point2, List<IEntity>> _gridContents = new HashMap<>();

    public static IEntity addObject(IEntity Entity) {
        Entity.loadContent();
        _contents.add(Entity);
        addToGrid(Entity);
        return Entity;
    }

    public static IEntity getObject(EntityType type) {
        if (_contents != null) {
            for (IEntity entity : _contents) {
                if (entity.getEntityType() == type) {
                    return entity;
                }
            }
        }
        return null;
    }

    private static final List<IEntity> gopResults = new ArrayList<>();

    public static List<IEntity> getEntities(EntityType type, Point2 target)

    {
        if (_contents != null) {
            goResults.clear();
            goResults.addAll(getObjects(type));
            gopResults.clear();
            for (IEntity goResult : goResults) {
                if (goResult.contains(target)) {
                    gopResults.add(goResult);
                }
            }
            return gopResults;
        }
        return null;
    }

    private static final List<IEntity> goResults = new ArrayList<>();

    public static List<IEntity> getObjects(EntityType type) {
        goResults.clear();
        for (IEntity _content : _contents) {
            if (_content.getEntityType() == type) {
                goResults.add(_content);
            }
        }
        return goResults;
    }

    private static final List<IActor> creatures = new ArrayList<>();

    public static List<IActor> getActors(ActorType type) {
        creatures.clear();
        if (type != ActorType.NONPLAYER) {
            for (IEntity elem : _contents) {
                if (elem.getEntityType() == EntityType.ACTOR) {
                    if (((IActor) elem).getActorType() == type) {
                        creatures.add(((IActor) elem));
                    }
                }
            }
        } else {
            for (IEntity elem : _contents) {
                if (elem.getEntityType() == EntityType.ACTOR) {
                    if (((IActor) elem).getActorType() != ActorType.PLAYER) {
                        creatures.add(((IActor) elem));
                    }
                }
            }
        }
        return creatures;
    }

    private static IActor _nextResult;

    public static List<IActor> getActorsAt(Point2 target, ActorType actorType) {
        creatures.clear();
        for (IEntity elem : _gridContents.get(target)) {
            if (elem.getEntityType() == EntityType.ACTOR) {
                _nextResult = (IActor) elem;
                if (actorType == null || _nextResult.getActorType() == actorType || (actorType == ActorType.NONPLAYER && _nextResult.getActorType() != ActorType.PLAYER)) {
                    creatures.add(_nextResult);
                }
            }
        }
        return creatures;
    }

    public static List<IActor> getActorsAt(Point2 target) {
        return getActorsAt(target, null);
    }

    private static final List<IActor> creatures2 = new ArrayList<>();

    public static List<IActor> getActorsSurrounding(Point2 target, int distance)

    {
        creatures2.clear();
        for (int ii = -distance; ii < distance + 1; ii++) {
            for (int jj = -distance; jj < distance + 1; jj++) {
                if (ii != 0 || jj != 0) {
                    for (IActor creature : getActorsAt(target.add(new Point2(ii, jj)), null)) {
                        creatures2.add(creature);
                    }
                }
            }
        }
        return creatures2;
    }

    public static boolean isLocationBlocked(Point2 location) {
        for (IEntity elem : _gridContents.get(location)) {
            if (elem.isBlocking()) {
                return true;
            }
        }
        return false;
    }

    private static List<IEntity> gotcResults;

    public static List<IEntity> getObjectsToCache() {
        gotcResults.clear();
        for (IEntity _content : _contents) {
            if (_content.getEntityType() != EntityType.FLOOR) {
                gotcResults.add(_content);
            }
        }
        return gotcResults;
    }

    public static IActor getNearestPlayer(IEntity target) {
        List<IActor> actors = getActors(ActorType.PLAYER);
        if (actors.size() > 0) {
            IEntity closest = actors.get(0);
            IEntity player;
            for (IActor actor : actors) {
                player = actor;
                if (HitTest.getDistanceSquare(target, player) < HitTest.getDistanceSquare(target, closest)) {
                    closest = player;
                }
            }
            return (IActor) closest;
        }
        return null;
    }

    public static IActor getNearestPlayer(IActor target)

    {
        return getNearestPlayer((IEntity) target);
    }

    public static void addObjects(List<IEntity> cache) {
        _contents.addAll(cache);
    }

    public static boolean anyContains(Point2 target, EntityType type) {
        for (IEntity entity : _gridContents.get(target)) {
            if (entity.getEntityType() == type) {
                return true;
            }
        }
        return false;
    }

    public static void removeObject(IEntity target) {
        _contents.remove(target);
    }

    public static void clear() {
        _contents = new ArrayList<>();
        _gridContents = new HashMap<>();
    }

    public static void reset() {
        _contents = new ArrayList<>();
        _gridContents = new HashMap<>();
        creatures.clear();
        loadContent();
    }

    public static void update() {
        for (int ii = 0; ii < _contents.size(); ii++) {
            if (ii >= _contents.size()) {
                return;
            }
            if (!_contents.get(ii).isActive()) {
                _gridContents.get(_contents.get(ii).getLocation()).remove(_contents.get(ii));
                _contents.remove(_contents.get(ii));
                ii--;
                continue;
            }
            _contents.get(ii).update();
        }
    }

    public static void draw() {
        if (SpxManager.Renderer != null) {
            for (IEntity component : _contents) {
                component.draw();
            }
        }
    }

    public static void loadContent() {
        if (SpxManager.Renderer != null) {
            for (IEntity component : _contents) {
                component.loadContent();
            }
        }
    }

    private static void addToGrid(IEntity entity) {
        if (!_gridContents.containsKey(entity.getLocation())) {
            _gridContents.put(entity.getLocation(), new ArrayList<IEntity>());
        }
        _gridContents.get(entity.getLocation()).add(entity);
    }

    public static void updateGridLocation(IEntity Entity, Point2 oldLocation) {
        if (_gridContents != null && oldLocation != null) {
            _gridContents.get(oldLocation).remove(Entity);
            addToGrid(Entity);
        }
    }

    private static final List<IActor> _players = new ArrayList<>();

    public static List<IActor> getPlayers() {
        _players.clear();
        for (IEntity tile : _contents) {
            if (tile.getEntityType() == EntityType.ACTOR && ((IActor) tile).getActorType() == ActorType.PLAYER) {
                _players.add((IActor) tile);
            }
        }
        return _players;
    }

    public static Point2 getEmptyLocation() {
        List<Point2> emptyLocations = new ArrayList<>();
        for (Point2 location : _gridContents.keySet()) {
            if (location.GridX > 0 && location.GridY > 0 && location.GridX < Settings.get().tileMapWidth - 1 && location.GridY < Settings.get().tileMapHeight - 1) {
                boolean exclude = false;
                for (int ii = 0; ii < _gridContents.get(location).size(); ii++) {
                    if (_gridContents.get(location).get(ii).getEntityType() == EntityType.ACTOR) {
                        exclude = true;
                    }
                }
                if (!exclude) {
                    emptyLocations.add(location);
                }
            }
        }
        return emptyLocations.get(RNG.next(0, emptyLocations.size()));
    }

    public static List<IEntity> getEntitiesToCache() {
        List<IEntity> results = new ArrayList<>();
        for (IEntity _content : _contents) {
            if (_content.getEntityType() != EntityType.FLOOR) {
                results.add(_content);
            }
        }
        return results;
    }

    public static IActor getTouchingCreature(IEntity entity) {
        for (IEntity _content : _contents) {
            if (_content.getEntityType() == EntityType.ACTOR) {
                if (_content.contains(entity.getLocation())) {
                    return (IActor) _content;
                }
            }
        }
        return null;
    }
}
